### PR TITLE
This refactors the dhcpd lease parser in the vmware builders and adds unit tests for everything.

### PR DESCRIPTION
**Pull request by arizvisa**
This PR re-impements the dhcpd lease parser that is being used by the vmware builder, and refactors the parsers that are used. Unit tests were written for all the parsers too.

What this does is pulls the logic from `builder/vmware/common/driver.go` that parses the dhcpd leases and replaces the regexes with an actual tokenizer that collects all of the entries within a dhcpd lease. Each entry in the lease is then exposed by a structure that can be iterated through and filtered. All fields within a lease entry are also converted into their native types. This logic is performed by new functionality `ReadDhcpdLeaseEntries` introduced in `builder/vmware/common/driver_parser.go`.

After this, `builder/vmware/common/driver.go`'s parsing logic is then replaced with a call to `ReadDhcpdLeaseEntries` which is then filtered according to the original logic. This PR actually intends to retain the original functionality whilst only refactoring the logic so that it's easier to build upon. Ideally it should work the exact same way.

This is the second part of solving issue #8528, and is based upon PR #9303.

It was broken up into 3 stages as the first stage (#9303) introduced unit tests to the parsers (which should definitely be merged).

The second stage (this one) refactors the logic of the dhcpd parsers and re-implements it so that it's backwards compatible.

The third stage (if possible) will actually change the way that the GuestIP is being fetched so that a list of leases can be checked.

### CODE REVIEW

**Comment by SwampDragons** [comment]
Hi, thanks for opening! I'm probably not going to get a chance to review closely until Monday but this looks good at a first glance.

**Comment by arizvisa** [comment]
Cool. No worries. If anything, the first part of this (PR #9303) should 100% be prioritized first, though, as it's the unit-tests for the parsers.

I needed those tests to ensure that this PR didn't break any prior functionality. So I'd review that those tests are actually comprehensive (and not too dumb)... because if they are, then this PR will likely break pre-existing functionality. I didn't really expect to finish this as quick as I did, so there's likely some oversights on my part (some of the commits were between 1am and 9am).

**Comment by arizvisa** [comment]
Oh. One other thing about this PR that it changes, is that the "sentinelSignaller" channel in the parsers (which was used as out-of-band signalling when running out of input) is disposed of. This changes the way that everything works, but simplifies things greatly as opposed to having to manage more than one channel when using any of the parsers.

**Comment by SwampDragons** [comment]
Thanks -- I meant to sit down with this today but didn't end up having enough time for the focus it deserves. It's first on my TODO list for tomorrow.

**Comment by arizvisa** [comment]
no worries. already thx for getting to them so quick.

**Comment by arizvisa** [comment]
Ok. I'll see what I can do.

File: builder/vmware/common/driver_parser.go [hunk 1]
```diff
@@ -1,117 +1,133 @@
...
-						continue
-					}
-					state += string(ch)
+		for {
+			by, ok := <-in
+			if !ok {
+				break
+			}
+
+			// If we're in a quote, then we continue until we're not in a quote
```
[the remark this answers is not in the document]
Comment (arizvisa) [reply builder/vmware/common/driver_parser.go:67]: Hah. That's y'alls infuence.

**Comment by arizvisa** [comment]
Ok. Got all your suggestions and added unit-tests for the networking file. Thanks for digging and noting the channel issue in your review. I realize it was probably a pain in the ass.

**Comment by arizvisa** [comment]
(although, hopefully the tests pass. heh.. i'm gonna sleep tho and if there's any further issues i'll try fix sometime in the afternoon when i wake up).

**Comment by SwampDragons** [comment]
It wasn't that bad actually :D  I knew right away there was something going on in the channels based on the changes you made, so it wasn't too hard to narrow down. Your new commit fixes it; I just ran a vmware smoke test locally and it passes.

**Pull request merged**